File: examples/anip-ts/src/sinks.ts
/**
 * Checkpoint sink interface and implementations.
 *
 * Mirrors the Python reference implementation's sinks.py.
 */

import { writeFileSync, mkdirSync } from "fs";
import { join } from "path";

export interface CheckpointSink {
  /** Publish a checkpoint. Should be idempotent. */
  publish(checkpoint: Record<string, unknown>): void;
}

/**
 * Writes checkpoints as JSON files to a local directory.
 *
 * Reference implementation — not a real external anchor.
 */
export class LocalFileSink implements CheckpointSink {
  private directory: string;

  constructor(directory: string) {
    this.directory = directory;
    mkdirSync(directory, { recursive: true });
  }

  publish(checkpoint: Record<string, unknown>): void {
    const filename = `${checkpoint.checkpoint_id}.json`;
    writeFileSync(
      join(this.directory, filename),
      JSON.stringify(checkpoint, Object.keys(checkpoint).sort(), 2),
    );
  }
}

File: examples/anip/anip_server/primitives/sinks.py
"""Checkpoint sink interface and implementations."""
import json
import os
from abc import ABC, abstractmethod
from typing import Any


class CheckpointSink(ABC):
    """Interface for publishing checkpoints to external storage."""

    @abstractmethod
    def publish(self, checkpoint: dict[str, Any]) -> None:
        """Publish a checkpoint. Should be idempotent."""
        ...


class LocalFileSink(CheckpointSink):
    """Writes checkpoints as JSON files to a local directory.

    Reference implementation — not a real external anchor.
    """

    def __init__(self, directory: str) -> None:
        self._directory = directory
        os.makedirs(directory, exist_ok=True)

    def publish(self, checkpoint: dict[str, Any]) -> None:
        filename = f"{checkpoint['checkpoint_id']}.json"
        path = os.path.join(self._directory, filename)
        with open(path, "w") as f:
            json.dump(checkpoint, f, indent=2, sort_keys=True)
